Updates to ADR-020 (protobuf signing) (#6359)
cosmos/cosmos-sdk: docs/architecture/adr-020-protobuf-transaction-encoding.md

package cosmos_sdk.v1;

message Tx {

>     // A protobuf serialization of a TxBody that matches representation in SignDoc.

Reviewer: Why does a `Tx` now have a `bytes` blob? Can we still achieve the desired outcome without this change? When I look at a tx, I would imagine I could understand and read all the fields.

Author: I tried to explain that in point 2. of the PR description, may be that helps. Anyways, let me try to go one step back to explain the underlying challenge. Maybe it is important to note that I'm not trying to change things here but cleanup the document based on what we agreed on.

A protobuf document is an abtract data structure that has a practically unlimited number of valid binary representation. Having a deterministic binary representation has never been a design goal of protobuf (like JSON and in contrast to other formarts like e.g. [Borsh](https://borsh.io/)). In order to verify signatures, the transaction signer and the verifier need to be able to agree on the exact same binary representation for the data. There are two ways to get there:
1. Specify a canonical representation (which is a custom addition to protobuf), that all implementations need to implement
2. Encode the protobuf document to bytes only once and maintain the binary representation

There is nothing wrong with 1. or 2., they just have different trade-offs. Aaron initially worked towards 1., but during the course of development explicitely discarded the idea of canonicalization which is documented in this file.

So we're going for 2., i.e. the signer chooses one binary representation and signs exactly that. But how is this binary representation transmitted to the verifier that only gets a `Tx`? An abspract data model (`Tx` with embedded `TxBody`) is not sufficient as it does not define the exact bytes needed for verification.

--------

Since the messages are [embedded as binary data because of the Any type](https://github.com/protocolbuffers/protobuf/blob/v3.12.3/src/google/protobuf/any.proto#L153-L154), you need a multi step decoding anyways to transform the binary in a complete fully nested tx model in memory. There are 3 main decoding steps to get there:
1. Decode raw data into `Tx`
2. Decode a binary body into `TxBody`
3. For each message decode the bytes in the `Any` value into the message type

This three step decoding is not the most simple thing on earth, but given the requirements that were expressed along the way, I think it has aceptable complexity and is very clean.

Reviewer: > But how is this binary representation transmitted to the verifier that only gets a Tx?

Same way it does it now. The verifier (i.e AnteHandler), computes the raw binary data and verified the sig. It takes what it needs to from the Tx and from the state (nonce and sequence) and verifies against the signature.

Author: > The verifier (i.e AnteHandler), computes the raw binary data and verified the sig. It takes what it needs to from the Tx and from the state (nonce and sequence) and verifies against the signature.

So this is implemented already? Could you please point me to the code you are referring to? I'm very unfamiliar with the codebase and was not very successful with full text search.

Reviewer: https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/ante/sigverify.go#L193-L214
https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/types/stdtx.go#L248-L260
https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/types/stdtx.go#L292-L313

Author: Thank you, this is very helpful. The last link (the implementation of `StdSignBytes`) shows that this implements the legazy JSON verification mode, not the direct protobuf verification. In legazy JSON there is a canonicalization step [sdk.MustSortJSON(bz), implemented here](https://github.com/cosmos/cosmos-sdk/blob/bd90d6b34448775e77b21676597c9fb42b1a65d3/types/utils.go#L24-L50).

In our discussions and also in noted in some places in ADR-020 it was decided to not require canonicalization of any protobuf struture other than the outermost `SignDoc`. SignDoc is simple enough to not require canonicalization other than _no extra data_ and _fields sorted by number_, which we probably get for free in all implementations.

We can challenge this decision, but we should be explicit about it.

message Tx {
    TxBody body = 1;
    AuthInfo auth_info = 2;
    // A list of signatures that matches the length and order of AuthInfo's signer_infos to
    // allow connecting signature meta information like public key and signing mode by position.
    repeated bytes signatures = 3;
}

// A variant of Tx that pins the signer's exact binary represenation of body and
// auth_info. This is used for signing, broadcasting and verification. The binary
// `serialize(tx: TxRaw)` is stored in Tendermint and the hash `sha256(serialize(tx: TxRaw))`
// becomes the "txhash", commonly used as the transaction ID.
message TxRaw {

>     // A protobuf serialization of a TxBody that matches representation in SignDoc.

Reviewer 2: ```suggestion
    // A protobuf serialization of a TxBody that matches the representation in SignDoc.
```

    bytes body = 1;

>     // A protobuf serialization of a AuthInfo that matches representation in SignDoc.

Reviewer 2: ```suggestion
    // A protobuf serialization of an AuthInfo that matches the representation in SignDoc.
```

    bytes auth_info = 2;
    // A list of signatures that matches the length and order of AuthInfo's signer_infos to
    // allow connecting signature meta information like public key and signing mode by position.
    repeated bytes signatures = 3;
}

message TxBody {
    // A list of messages to be executed. The required signers of those messages define
    // the number and order of elements in AuthInfo's signer_infos and Tx's signatures.
    // Each required signer address is added to the list only the first time it occurs.
    //
    // By convention, the first required signer (usually from the first message) is referred
    // to as the primary signer and pays the fee for the whole transaction.
    repeated google.protobuf.Any messages = 1;
    string memo = 2;
    int64 timeout_height = 3;
    repeated google.protobuf.Any extension_options = 1023;
}

message AuthInfo {
    // This list defines the signing modes for the required signers. The number
    // and order of elements must match the required signers from TxBody's messages.
    // The first element is the primary signer and the one which pays the fee.
    repeated SignerInfo signer_infos = 1;

>     // The fee can be calculated by looking into the body and the signer infos.

Reviewer 2: ```suggestion
    // The fee can be calculated based on the cost of evaluating the body and doing signature verification base of the signers. This can be estimated via simulation.
```

Author: What is a "signature verification base"? Is this a typo in the suggestion?

Reviewer 2: Oops! Updated the suggestion

subtle differences between the signing and encoding formats which could 
potentially be exploited by an attacker)

Signatures are structured using the `SignDoc` below which reuses the serialization of
`TxBody` and `AuthInfo` and only adds the fields which are needed for signatures:

```proto
// types/types.proto
message SignDoc {

>     // A protobuf serialization of a TxBody that matches representation in TxRaw.

Reviewer 2: ```suggestion
    // A protobuf serialization of a TxBody that matches the representation in TxRaw.
```

    bytes body = 1;

>     // A protobuf serialization of a AuthInfo that matches representation in TxRaw.

Reviewer 2: ```suggestion
    // A protobuf serialization of an AuthInfo that matches the representation in TxRaw.
```

In order to sign in the default mode, clients take the following steps:

> 1. Serialize `TxBody` and `AuthInfo` using any valid protobuf serialization

Reviewer 2: ```suggestion
1. Serialize `TxBody` and `AuthInfo` using any valid protobuf implementation.
```

Author: Hah, good one. I actually wanted to say "using any valid protobuf serialization" as in output of a serializer, but this does not work well with the beginning of the sentence. Will take your version.

2. Create a `SignDoc` and encode it. (The only requirement of the underlying
   protobuf implementation is that fields are serialized in order).

> 3. Sign the encoded `SignDoc` bytes
> 4. Build a `TxRaw` and serialize it for broadcasting

Reviewer 2: ```suggestion
3. Sign the encoded `SignDoc` bytes.
4. Build a `TxRaw` and serialize it for broadcasting.
```

Signature verification is based on comparing the raw `TxBody` and `AuthInfo`
bytes encoded in `TxRaw` not based on any ["canonicalization"](https://github.com/regen-network/canonical-proto3)
algorithm which creates added complexity for clients in addition to preventing
some forms of upgradeability (to be addressed later in this document).

Signature verifiers do:

> 1. Pull out `body` and `auth_info` from a `TxRaw` and deserialize them
> 3. Create a list of required signer addresses from the messages

Reviewer 2: ```suggestion
1. Deserialize a `TxRaw` and pull out `body` and `auth_info`.
2. Create a list of required signer addresses from the messages.
```
